src/back/spirv/layout.rs

```diff
@@ -0,0 +1,216 @@
+            global_variables: vec![],
+            function_declarations: vec![],
+            function_definitions: vec![],
+        }
+    }
+
+    pub(crate) fn in_words(&self) -> Vec<u32> {
+        let mut words: Vec<Word> = vec![];
+
+        for capability in self.capabilities.iter() {
+            words.push(*capability);
+        }
+
+        for extension in self.extensions.iter() {
+            words.push(*extension);
+        }
+
+        for ext_inst_import in self.ext_inst_imports.iter() {
+            words.push(*ext_inst_import);
+        }
+
+        for memory_model in self.memory_model.iter() {
+            words.push(*memory_model);
+        }
+
+        for entry_point in self.entry_points.iter() {
+            words.push(*entry_point);
+        }
+
+        for execution_mode in self.execution_modes.iter() {
+            words.push(*execution_mode);
+        }
+
+        for debug in self.debugs.iter() {
+            words.push(*debug);
+        }
+
+        for annotation in self.annotations.iter() {
+            words.push(*annotation);
+        }
+
+        for type_declaration in self.type_declarations.iter() {
+            words.push(*type_declaration);
+        }
+
+        for constants in self.constants.iter() {
+            words.push(*constants);
+        }
+
+        for global_variables in self.global_variables.iter() {
+            words.push(*global_variables);
+        }
+
+        for function_declaration in self.function_declarations.iter() {
+            words.push(*function_declaration);
+        }
+
+        for function_definition in self.function_definitions.iter() {
+            words.push(*function_definition);
+        }
+
+        words
+    }
+}
+
+pub(crate) struct Instruction {
+    op: u32,
+    wc: u32,
+    type_id: Option<Word>,
+    result_id: Option<Word>,
+    operands: Vec<Word>,
+}
+
+impl Instruction {
+    pub(crate) fn new(op: Op) -> Self {
+        Instruction {
+            op: op as u32,
+            wc: 1, // Always start at 1 for the first word (OP + WC),
+            type_id: None,
+            result_id: None,
+            operands: vec![],
+        }
+    }
+
+    pub(crate) fn set_type(&mut self, id: Word) {
+        if self.type_id.is_none() {
+            self.wc += 1;
+        }
+        self.type_id = Some(id);
+    }
+
+    pub(crate) fn set_result(&mut self, id: Word) {
+        if self.result_id.is_none() {
+            self.wc += 1;
+        }
+        self.result_id = Some(id);
+    }
+
+    pub(crate) fn add_operand(&mut self, operand: Word) {
+        self.operands.push(operand);
+        self.wc += 1;
+    }
+
+    pub(crate) fn add_operands(&mut self, operands: Vec<Word>) {
+        for operand in operands {
+            self.add_operand(operand);
+        }
+    }
+
+    pub(crate) fn to_words(&self) -> Vec<Word> {
+        let mut words = Vec::with_capacity(self.wc as usize);
+        let wc_op = (self.wc << 16 | self.op) as u32;
+
+        words.push(wc_op);
```

Comment:
putting the whole problem of heap allocations aside, all of this can be simplified like:
```rust
{
iter::once((self.wc << 16 | self.op) as u32)
  .chain(self.type_id)
  .chain(self.result_id)
  .chain(self.operands)
}
```

Reply:
I can do this:
```rust
std::iter::once((self.wc << 16 | self.op as u32) as u32)
    .chain(self.type_id)
    .chain(self.result_id)
    .chain(self.operands)
    .collect::<Vec<_>>()
```
But I get the error that self.operands cannot be moved (because Vec does not have the Copy trait). Is there an easy way to fix this. If that works, then I really like this solution.

Reply:
right, in order to avoid moving `operands` out, we need to iterate it and clone the items: `self.operands.iter().cloned()`